pkg/infoschema/builder.go

```diff
@@ -341,6 +341,18 @@ func (b *Builder) getTableIDs(diff *model.SchemaDiff) (oldTableID, newTableID in
 		newTableID = diff.TableID
 	case model.ActionDropTable, model.ActionDropView, model.ActionDropSequence:
 		oldTableID = diff.TableID
+
+		// Still keep the table in infoschema until when the state of table reaches StateNone. This is because
+		// the `ON DELETE/UPDATE CASCADE` function of foreign key may need to find the table in infoschema. Not
+		// only the table with foreign keys are stored in the infoschema to keep a consistent behavior for all
+		// tables.
+		tblInfo, err := m.GetTable(diff.SchemaID, oldTableID)
```

Comment:
Here is the key change to avoid removing `WriteOnly` table from the infoschema.

Reply:
FYI: In the past, there is a bug https://github.com/pingcap/tidb/issues/54906 and the behaviour is:
infoschema full load do not load any non-public table
apply diff load the tables in the intermediate state.

And it's fixed to load all non-publice tables recently.